lib/adapter.js

```diff
@@ -5647,47 +5647,51 @@ function Adapter(options) {
         clearInterval(reportInterval);
         const id = 'system.adapter.' + that.namespace;
 
+        function finishUnload() {
+            if (that.states) {
+                that.outputCount++;
+                that.states.setState(id + '.alive', {val: false, ack: true, from: id}, () => {
+                    if (!isPause && that.log) logger.info('terminating');
+                    that.terminate(isScheduled ? EXIT_CODES.START_IMMEDIATELY_AFTER_STOP: 0);
+                });
+            }
+        }
+
         if (typeof options.unload === 'function') {
-            options.unload(() => {
-                if (that.states) {
-                    that.outputCount++;
-                    that.states.setState(id + '.alive', {val: false, ack: true, from: id}, () => {
-                        if (!isPause && that.log) logger.info('terminating');
-                        that.terminate(isScheduled ? EXIT_CODES.START_IMMEDIATELY_AFTER_STOP: 0);
-                    });
+            if (options.unload.length >= 1) {
+                // The method takes (at least) a callback
+                options.unload(finishUnload);
+            } else {
+                // The method takes no arguments, so it must return a Promise
+                const unloadPromise = options.unload();
+                if (unloadPromise instanceof Promise) {
+                    // Call finishUnload in the case of success and failure
+                    unloadPromise.then(finishUnload, finishUnload);
+                } else {
+                    // No callback accepted and no Promise returned - force unload
+                    logger.error(`Error in ${id}: The unload method must return a Promise if it does not accept a callback!`);
                 }
-            });
+            }
         } else {
-            that.emit('unload', () => {
-                if (that.states) {
-                    that.outputCount++;
-                    that.states.setState(id + '.alive', {val: false, ack: true, from: id}, () => {
-                        if (!isPause && that.log) logger.info('terminating');
-                        that.terminate(isScheduled ? EXIT_CODES.START_IMMEDIATELY_AFTER_STOP: 0);
-                    });
-                }
-            });
+            that.emit('unload', finishUnload);
+        }
 
-            // Make delay to let event 'unload' to be processed
-            setTimeout(() => {
-                if (that.states) {
-                    that.outputCount++;
-                    that.states.setState(id + '.alive', {val: false, ack: true, from: id}, () => {
-                        if (!isPause && that.log) logger.info('terminating');
-                        that.terminate(isScheduled ? EXIT_CODES.START_IMMEDIATELY_AFTER_STOP: 0);
-                    });
+        // Even if the developer forgets to call the unload callback, we need to stop the process
+        // Therefore wait a short while and then force the unload
+        setTimeout(() => {

```

Comment:
This used to be in the `else` branch and therefore was only called if the `unload` event was used, but not when `unload` was a function on the options object.

But from my understanding, the adapter process should be force-terminated in both cases, shouldn't it? After all, there's no difference where the developer forgets to call the callback.

@GermanBluefox @Apollon77

Reply:
I see it the same